### PR TITLE
Persist model definition in model metadata

### PR DESCRIPTION
**Pull request by ryanbogan**
### Description
Adds MethodComponentContext to ModelMetadata.  This allows developers and users to more easily access the model definition after the model has been created.
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/indices/ModelMetadata.java [hunk 1]
```diff
@@ -21,22 +21,17 @@
...
 
-import static org.opensearch.knn.common.KNNConstants.DIMENSION;
-import static org.opensearch.knn.common.KNNConstants.KNN_ENGINE;
-import static org.opensearch.knn.common.KNNConstants.METHOD_PARAMETER_SPACE_TYPE;
-import static org.opensearch.knn.common.KNNConstants.MODEL_DESCRIPTION;
-import static org.opensearch.knn.common.KNNConstants.MODEL_ERROR;
-import static org.opensearch.knn.common.KNNConstants.MODEL_STATE;
-import static org.opensearch.knn.common.KNNConstants.MODEL_TIMESTAMP;
-import static org.opensearch.knn.common.KNNConstants.MODEL_NODE_ASSIGNMENT;
+import static org.opensearch.knn.common.KNNConstants.*;
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/indices/ModelMetadata.java:34]: I prefer to avoid wildcard imports. lets avoid

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 2]
```diff
@@ -193,6 +197,69 @@ public Map<String, Object> getParameters() {
...
+                String value = entry.getValue().toString();
+                value = value.replace(",", "$%$");
+                stringBuilder.append(value).append(";");
+            }
+        }
+        stringBuilder.append("]}");
+        return stringBuilder.toString();
+    }
+
+    public static MethodComponentContext fromString(String in) {
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:217]: This method will need a few comments describing what it will do. Also, lets add java doc

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 3]
```diff
@@ -193,6 +197,69 @@ public Map<String, Object> getParameters() {
...
     }
 
+    @Override
+    public String toString() {
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append("{name=").append(name).append(";");
+        stringBuilder.append("parameters=[");
+        if (Objects.nonNull(parameters)) {
+            for (Map.Entry<String, Object> entry : parameters.entrySet()) {
+                stringBuilder.append(entry.getKey()).append("=");
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:207]: Can you add a comment that these will always be either strings, ints, or methodcomponentcontexts and reference the parameter.java file?

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 4]
```diff
@@ -193,6 +197,69 @@ public Map<String, Object> getParameters() {
...
+            }
+        }
+        stringBuilder.append("]}");
+        return stringBuilder.toString();
+    }
+
+    public static MethodComponentContext fromString(String in) {
+        int index = 0;
+
+        String[] outerMethodComponentContextArray = in.split("\\{", -1);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:220]: Why do we need "\\"?

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 5]
```diff
@@ -193,6 +197,69 @@ public Map<String, Object> getParameters() {
...
+                String stringValue = substring.substring(substring.indexOf("=") + 1);
+                Object value;
+                if (stringValue.isEmpty()) {
+                    value = fromString(outerMethodComponentContextArray[index]);
+                } else if (NumberUtils.isNumber(stringValue)) {
+                    value = Integer.parseInt(stringValue);
+                } else if (stringValue.equals("true") || stringValue.equals("false")) {
+                    value = Boolean.parseBoolean(stringValue);
+                } else {
+                    value = stringValue;
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:251]: Can we add back commas?

**Comment by jmazanec15** [comment]
@ryanbogan In PR description, can you add longer description of the change?

**Comment by jmazanec15** [comment]
Also @ryanbogan can we update https://github.com/opensearch-project/k-NN/blob/main/src/main/resources/mappings/model-index.json

File: src/main/java/org/opensearch/knn/index/KNNMethodContext.java [hunk 6]
```diff
@@ -65,6 +65,8 @@ public static synchronized KNNMethodContext getDefault() {
     @NonNull
     private final MethodComponentContext methodComponentContext;
 
+    private static final String DELIMITER = ";";
```
[the remark this answers is not in the document]
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/index/KNNMethodContext.java:68]: Nope, that was leftover from an earlier iteration of the code.

File: src/main/java/org/opensearch/knn/indices/ModelMetadata.java (hunk 1 above)
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/indices/ModelMetadata.java:34]: The IDE changed it automatically, will remove

File: src/main/java/org/opensearch/knn/indices/ModelMetadata.java [hunk 7]
```diff
@@ -334,7 +399,8 @@ public static ModelMetadata getMetadataFromSourceMap(final Map<String, Object> m
             objectToString(timestamp),
             objectToString(description),
             objectToString(error),
-            objectToString(trainingNodeAssignment)
+            objectToString(trainingNodeAssignment),
+            (MethodComponentContext) methodComponentContext
```
[the remark this answers is not in the document]
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/indices/ModelMetadata.java:403]: Yes, otherwise it throws a parsing exception

File: src/test/java/org/opensearch/knn/index/codec/KNNCodecTestCase.java [hunk 8]
```diff
@@ -53,11 +53,7 @@
 import java.io.IOException;
 import java.time.ZoneOffset;
 import java.time.ZonedDateTime;
-import java.util.Arrays;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
[the remark this answers is not in the document]
Comment (ryanbogan) [reply src/test/java/org/opensearch/knn/index/codec/KNNCodecTestCase.java:56]: same as above, will remove

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java (hunk 4 above)
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/index/MethodComponentContext.java:220]: The split method takes a regex as an input.  { is a special character in regex so the \\ escapes it to just split by {

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 9]
```diff
@@ -193,6 +197,69 @@ public Map<String, Object> getParameters() {
...
+        stringBuilder.append("]}");
+        return stringBuilder.toString();
+    }
+
+    public static MethodComponentContext fromString(String in) {
+        int index = 0;
+
+        String[] outerMethodComponentContextArray = in.split("\\{", -1);
+
+        if (outerMethodComponentContextArray[index].isEmpty()) {
```
[the remark this answers is not in the document]
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/index/MethodComponentContext.java:222]: Sure, I'll try to see if it works.

[the remark this answers is not in the document]
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/index/MethodComponentContext.java:222]: Added a parse parameters method.  Didn't add a parse name method because it's only two lines

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java (hunk 2 above)
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/index/MethodComponentContext.java:217]: Added a java doc with a small description.  Let me know if I need to add more

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 10]
```diff
@@ -193,6 +200,92 @@ public Map<String, Object> getParameters() {
...
+    @Override
+    public String toString() {
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append("{name=").append(name).append(DELIMITER);
+        stringBuilder.append("parameters=[");
+        if (Objects.nonNull(parameters)) {
+            for (Map.Entry<String, Object> entry : parameters.entrySet()) {
+                stringBuilder.append(entry.getKey()).append("=");
+                String value = entry.getValue().toString();
+                // Model Metadata uses a delimiter to split the input string in its fromString method
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:220]: Link to ModelMetadata code and line number for reference (2.12 tag). This will allow future devs to see exactly where this is happening

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 11]
```diff
@@ -193,6 +200,92 @@ public Map<String, Object> getParameters() {
...
+                // If any of the values in the method component context contain this delimiter,
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:221]: Link to this issue: https://github.com/opensearch-project/k-NN/issues/1337

File: src/main/java/org/opensearch/knn/indices/ModelMetadata.java [hunk 12]
```diff
@@ -321,11 +379,16 @@ public static ModelMetadata getMetadataFromSourceMap(final Map<String, Object> m
         Object description = modelSourceMap.get(KNNConstants.MODEL_DESCRIPTION);
         Object error = modelSourceMap.get(KNNConstants.MODEL_ERROR);
         Object trainingNodeAssignment = modelSourceMap.get(KNNConstants.MODEL_NODE_ASSIGNMENT);
+        Object methodComponentContext = modelSourceMap.get(KNNConstants.MODEL_METHOD_COMPONENT_CONTEXT);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/indices/ModelMetadata.java:382]: I dont think this will work. Because we are indexing the methodComponentContext as a string, we are going it will be represented as a string in the source map. We should call from string on this to get the correct value.

Add a test for this as well.

File: src/test/java/org/opensearch/knn/indices/ModelMetadataTests.java [hunk 13]
```diff
@@ -14,13 +14,15 @@
 import org.opensearch.common.io.stream.BytesStreamOutput;
```
Comment (jmazanec15) [inline src/test/java/org/opensearch/knn/indices/ModelMetadataTests.java:14]: Can we add a couple dedicated tests in here when ModelContext is not empty?

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java (hunk 4 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/MethodComponentContext.java:241]: Can we avoid this by just confirming the first and last character are {} and then building a substring from then which is processed.

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 14]
```diff
@@ -193,6 +200,92 @@ public Map<String, Object> getParameters() {
         return parameters;
     }
 
+    /**
+     *
+     * Provides a String representation of MethodComponentContext
+     * Sample return:
+     * {name=ivf;parameters=[nlist=4;type=fp16;encoder={name=sq;parameters=[nprobes=2;clip=false;]};]}
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:207]: why do we need { and } at beginning and end? can we not skip it?

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 15]
```diff
@@ -193,6 +200,92 @@ public Map<String, Object> getParameters() {
...
+                // Parameters will always be a MethodComponentContext, String, integer, or boolean
+                // https://github.com/opensearch-project/k-NN/blob/2.12/src/main/java/org/opensearch/knn/index/Parameter.java
+                if (stringValue.isEmpty()) {
+                    value = fromString(outerMethodComponentContextArray[index]);
+                } else if (NumberUtils.isNumber(stringValue)) {
+                    value = Integer.parseInt(stringValue);
+                } else if (stringValue.equals("true") || stringValue.equals("false")) {
+                    value = Boolean.parseBoolean(stringValue);
+                } else {
+                    stringValue = stringValue.replace("$%$", ModelMetadata.DELIMITER);
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:277]: create a constant for $%$

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java (hunk 14 above)
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/index/MethodComponentContext.java:207]: I added those to help with the recursion if there is a nested MethodComponentContext object.  Otherwise, there is no way to differentiate the object from a normal field.

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java (hunk 4 above)
Comment (ryanbogan) [reply src/main/java/org/opensearch/knn/index/MethodComponentContext.java:241]: The split still needs to be there for the recursion to work.  Otherwise, I can't pass the whole substring of a nested MethodComponentContext because it will be split by ;

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 16]
```diff
@@ -193,6 +211,102 @@ public Map<String, Object> getParameters() {
...
+                // https://github.com/opensearch-project/k-NN/issues/1337
+                value = value.replace(ModelMetadata.DELIMITER, DELIMITER_PLACEHOLDER);
+                stringBuilder.append(value).append(DELIMITER);
+            }
+        }
+        stringBuilder.append("]}");
+        return stringBuilder.toString();
+    }
+
+    /**
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:250]: In the current logic, it is a bit hard to understand the parsing. Can we do something like the below instead?

```java
    public static MethodComponentContext fromString(String in) {
        String stringToParse = unwrapString(in, '{', '}');

        // Parse name from string
        String[] nameAndParameters = stringToParse.split(DELIMITER, 1);
        checkExpectedArrayLength(nameAndParameters, 2);
        String name = parseName(nameAndParameters[0]);
        String parametersString = nameAndParameters[1];
        Map<String, Object> parameters = parseParameters(parametersString);
        return new MethodComponentContext(name, parameters);
    }

    private static String parseName(String candidateNameString) {
        // Expecting candidateNameString to look like "name=ivf"
        checkStringNotEmpty(candidateNameString);
        String[] nameKeyAndValue = candidateNameString.split("=");
        checkExpectedArrayLength(nameKeyAndValue, 2);
        checkStringMatches(nameKeyAndValue[0], "name");
        return nameKeyAndValue[0];
    }

    private static Map<String, Object> parseParameters(String candidateParameterString) {
        checkStringNotEmpty(candidateParameterString);
        String[] parametersKeyAndValue = candidateParameterString.split("=", 1);
        checkExpectedArrayLength(parametersKeyAndValue, 2);
        checkStringMatches(parametersKeyAndValue[0], "parameters");
        return parseParametersValue(parametersKeyAndValue[1]);
    }

    private static Map<String, Object> parseParametersValue(String candidateParameterValueString) {
        // Expected input is [nlist=4;type=fp16;encoder={name=sq;parameters=[nprobes=2;clip=false;]};]
        checkStringNotEmpty(candidateParameterValueString);
        candidateParameterValueString = unwrapString(candidateParameterValueString, '[', ']');
        Map<String, Object> parameters = new HashMap<>();
        while (!candidateParameterValueString.isEmpty()) {
            String[] keyAndValueToParse = candidateParameterValueString.split("=", 1);
            String key = keyAndValueToParse[0];
            ValueAndRestToParse parsed = parseParameterValueAndRestToParse(keyAndValueToParse[1]);
            parameters.put(key, parsed.getValue());
            candidateParameterValueString = parsed.getRestToParse();
        }

        return parameters;
    }

    private static ValueAndRestToParse parseParameterValueAndRestToParse(String candidateParameterValueAndRestToParse) {
        if (candidateParameterValueAndRestToParse.charAt(0) == '{') {
            int endOfNestedMap = findClosingPosition(candidateParameterValueAndRestToParse, '{', '}');
            String nestedMethodContext = candidateParameterValueAndRestToParse.substring(0, endOfNestedMap);
            Object nestedParse = fromString(nestedMethodContext);
            String restToParse = candidateParameterValueAndRestToParse.substring(endOfNestedMap + 1);
            return new ValueAndRestToParse(nestedParse, restToParse);
        }

        String[] stringValueAndRestToParse = candidateParameterValueAndRestToParse.split(DELIMITER, 1);
        String stringValue = stringValueAndRestToParse[0];
        Object value;
        if (NumberUtils.isNumber(stringValue)) {
            value = Integer.parseInt(stringValue);
        } else if (stringValue.equals("true") || stringValue.equals("false")) {
            value = Boolean.parseBoolean(stringValue);
        } else {
            stringValue = stringValue.replace(DELIMITER_PLACEHOLDER, ModelMetadata.DELIMITER);
            value = stringValue;
        }

        return new ValueAndRestToParse(value, stringValueAndRestToParse[1]);
    }

    private static String unwrapString(String in, char expectedStart, char expectedEnd) {
        if (in.length() < 2) {
            throw new IllegalArgumentException("Invalid string.");
        }

        if (in.charAt(0) != expectedStart || in.charAt(in.length() - 1) != expectedEnd) {
            throw new IllegalArgumentException("Invalid string.");
        }
        return in.substring(1, in.length() - 1);
    }

    private static int findClosingPosition(String in, char expectedStart, char expectedEnd) {
        int nestedLevel = 0;
        for (int i = 0; i < in.length(); i++) {
            if (in.charAt(i) == expectedStart) {
                nestedLevel++;
                continue;
            }

            if (in.charAt(i) == expectedEnd) {
                nestedLevel--;
            }
            
            if (nestedLevel == 0) {
                return i;
            }
        }
        
        throw new IllegalArgumentException("Invalid string. No end to the nesting");
    }

    private static void checkStringNotEmpty(String string) {
        if (string.isEmpty()) {
            //TODO: Come up with better exception
            throw new RuntimeException("think of better exception");
        }
    }

    private static void checkStringMatches(String string, String expected) {
        if (!Objects.equals(string, expected)) {
            //TODO: Come up with better exception
            throw new RuntimeException("think of better exception");
        }
    }

    private static void checkExpectedArrayLength(String[] array, int expectedLength) {
        if (null == array) {
            //TODO: Come up with better exception
            throw new RuntimeException("think of better exception");
        }

        if (array.length != expectedLength) {
            //TODO: Come up with better exception
            throw new RuntimeException("think of better exception");
        }
    }

    @AllArgsConstructor
    @Getter
    private static class ValueAndRestToParse {
        private final Object value;
        private final String restToParse;
    }
```

File: src/main/java/org/opensearch/knn/indices/ModelMetadata.java [hunk 17]
```diff
@@ -76,6 +74,12 @@ public ModelMetadata(StreamInput in) throws IOException {
         } else {
             this.trainingNodeAssignment = "";
         }
+
+        if (IndexUtil.isVersionOnOrAfterMinRequiredVersion(in.getVersion(), IndexUtil.MODEL_METHOD_COMPONENT_CONTEXT_KEY)) {
+            this.methodComponentContext = new MethodComponentContext(in);
+        } else {
+            this.methodComponentContext = MethodComponentContext.DEFAULT;
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/indices/ModelMetadata.java:81]: Can we set this to null? It will be cleaner than checking if the name is empty to see if it is supported.

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 18]
```diff
@@ -11,24 +11,26 @@
...
 import org.opensearch.core.xcontent.ToXContentFragment;
 import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
 import org.opensearch.index.mapper.MapperParsingException;
 
 import java.io.IOException;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.*;
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:29]: Can we get rid of wildcard import?

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 19]
```diff
@@ -193,6 +211,196 @@ public Map<String, Object> getParameters() {
...
+            }
+        }
+
+        throw new IllegalArgumentException("Invalid string. No end to the nesting");
+    }
+
+    private static void checkStringNotEmpty(String string) {
+        if (string.isEmpty()) {
+            // TODO: Come up with better exception
+            throw new RuntimeException("think of better exception");
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:374]: Can you provide better exception? Lets throw IllegalArgException and remove TODOs

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 20]
```diff
@@ -193,6 +215,194 @@ public Map<String, Object> getParameters() {
...
+    private static void checkStringMatches(String string, String expected) {
+        if (!Objects.equals(string, expected)) {
+            throw new RuntimeException("Unexpected key in MethodComponentContext.  Expected 'name' or 'parameters'");
+        }
+    }
+
+    private static void checkExpectedArrayLength(String[] array, int expectedLength) {
+        if (null == array) {
+            // TODO: Come up with better exception
+            throw new RuntimeException("think of better exception");
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:390]: Fix exception. Switch to illegalargexception

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 21]
```diff
@@ -193,6 +215,194 @@ public Map<String, Object> getParameters() {
...
+        }
+
+        if (array.length != expectedLength) {
+            // TODO: Come up with better exception
+            throw new RuntimeException("not expected length");
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:395]: Fix exception

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 22]
```diff
@@ -193,6 +215,194 @@ public Map<String, Object> getParameters() {
...
+
+    private static void checkStringNotEmpty(String string) {
+        if (string.isEmpty()) {
+            throw new RuntimeException("Unable to parse MethodComponentContext");
+        }
+    }
+
+    private static void checkStringMatches(String string, String expected) {
+        if (!Objects.equals(string, expected)) {
+            throw new RuntimeException("Unexpected key in MethodComponentContext.  Expected 'name' or 'parameters'");
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:383]: Switch to IllegalArgException

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 23]
```diff
@@ -193,6 +215,194 @@ public Map<String, Object> getParameters() {
...
+        }
+
+        return parameters;
+    }
+
+    private static ValueAndRestToParse parseParameterValueAndRestToParse(String candidateParameterValueAndRestToParse) {
+        if (candidateParameterValueAndRestToParse.charAt(0) == '{') {
+            int endOfNestedMap = findClosingPosition(candidateParameterValueAndRestToParse, '{', '}');
+            String nestedMethodContext = candidateParameterValueAndRestToParse.substring(0, endOfNestedMap + 1);
+            System.out.println("nested method context: " + nestedMethodContext);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:322]: Remove

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 24]
```diff
@@ -193,6 +215,194 @@ public Map<String, Object> getParameters() {
...
+     */
+    public static MethodComponentContext fromClusterStateString(String in) {
+        String stringToParse = unwrapString(in, '{', '}');
+
+        // Parse name from string
+        String[] nameAndParameters = stringToParse.split(DELIMITER, 2);
+        System.out.println("nameAndParameters: " + Arrays.toString(nameAndParameters));
+        checkExpectedArrayLength(nameAndParameters, 2);
+        String name = parseName(nameAndParameters[0]);
+        System.out.println("name: " + name);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:269]: Remove

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 25]
```diff
@@ -193,6 +215,194 @@ public Map<String, Object> getParameters() {
...
+     *
+     * @param in a string representation of MethodComponentContext
+     * @return a MethodComponentContext object
+     */
+    public static MethodComponentContext fromClusterStateString(String in) {
+        String stringToParse = unwrapString(in, '{', '}');
+
+        // Parse name from string
+        String[] nameAndParameters = stringToParse.split(DELIMITER, 2);
+        System.out.println("nameAndParameters: " + Arrays.toString(nameAndParameters));
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:266]: Remove

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 26]
```diff
@@ -193,6 +215,194 @@ public Map<String, Object> getParameters() {
...
+        if (parametersKeyAndValue.length == 1) {
+            return Collections.emptyMap();
+        }
+        checkExpectedArrayLength(parametersKeyAndValue, 2);
+        return parseParametersValue(parametersKeyAndValue[1]);
+    }
+
+    private static Map<String, Object> parseParametersValue(String candidateParameterValueString) {
+        // Expected input is [nlist=4;type=fp16;encoder={name=sq;parameters=[nprobes=2;clip=false;]};]
+        System.out.println("parameter value: " + candidateParameterValueString);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:300]: Remove

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 27]
```diff
@@ -193,6 +215,194 @@ public Map<String, Object> getParameters() {
...
+            String nestedMethodContext = candidateParameterValueAndRestToParse.substring(0, endOfNestedMap + 1);
+            System.out.println("nested method context: " + nestedMethodContext);
+            Object nestedParse = fromClusterStateString(nestedMethodContext);
+            String restToParse = candidateParameterValueAndRestToParse.substring(endOfNestedMap + 1);
+            return new ValueAndRestToParse(nestedParse, restToParse);
+        }
+
+        String[] stringValueAndRestToParse = candidateParameterValueAndRestToParse.split(DELIMITER, 2);
+        String stringValue = stringValueAndRestToParse[0];
+        System.out.println("string value: " + stringValue);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:330]: Remove

File: src/main/java/org/opensearch/knn/index/MethodComponentContext.java [hunk 28]
```diff
@@ -193,6 +215,194 @@ public Map<String, Object> getParameters() {
...
+                return i;
+            }
+        }
+
+        throw new IllegalArgumentException("Invalid string. No end to the nesting");
+    }
+
+    private static void checkStringNotEmpty(String string) {
+        if (string.isEmpty()) {
+            throw new RuntimeException("Unable to parse MethodComponentContext");
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/MethodComponentContext.java:377]: IllegalArgException

File: src/main/java/org/opensearch/knn/indices/ModelMetadata.java [hunk 29]
```diff
@@ -321,11 +384,29 @@ public static ModelMetadata getMetadataFromSourceMap(final Map<String, Object> m
...
+                );
+                methodComponentContext = MethodComponentContext.fromXContent(xContentParser);
+            } catch (IOException e) {
+                throw new IllegalArgumentException("Error parsing method component context");
+            }
+        } else {
+            methodComponentContext = MethodComponentContext.EMPTY;
+        }
+
+        // {"name":"ivf","parameters":{"encoder":{"name":"pq","parameters":{"code_size":8,"m":2}}}}
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/indices/ModelMetadata.java:408]: Remove this

**Review by jmazanec15 — APPROVED** [review]
lgtm thanks!

**Review by VijayanB — APPROVED** [review]
